waggle-dance-extensions/src/test/java/com/hotels/bdp/waggledance/extensions/client/ratelimit/RateLimitingInvocationHandlerTest.java

```diff
@@ -0,0 +1,132 @@
+import com.hotels.bdp.waggledance.extensions.client.ratelimit.memory.InMemoryBucketService;
+import com.hotels.bdp.waggledance.server.WaggleDanceServerException;
+
+@RunWith(MockitoJUnitRunner.class)
+public class RateLimitingInvocationHandlerTest {
+
+  private static final String USER = "user";
+  private @Mock ThriftClientFactory thriftClientFactory;
+  private @Mock CloseableThriftHiveMetastoreIface client;
+  private @Mock BucketKeyGenerator bucketKeyGenerator;
+  private BucketService bucketService = new InMemoryBucketService(new IntervallyBandwidthProvider(2, 1));
+  private AbstractMetaStore metastore = AbstractMetaStore.newPrimaryInstance("name", "uri");
+  private CloseableThriftHiveMetastoreIface handlerProxy;
+
+  @Before
+  public void setUp() {
+    when(thriftClientFactory.newInstance(metastore)).thenReturn(client);
+    when(bucketKeyGenerator.generateKey(USER)).thenReturn(USER);
+    when(bucketKeyGenerator.generateKey(UNKNOWN_USER)).thenReturn(UNKNOWN_USER);
+    handlerProxy = new RateLimitingClientFactory(thriftClientFactory, bucketService, bucketKeyGenerator)
+        .newInstance(metastore);
+  }
+
+  @Test
+  public void testLimitDifferentUsers() throws Exception {
+
+    assertTokens(2, 2);
+    handlerProxy.get_table("db", "table");
+    assertTokens(2, 1);
+
+    handlerProxy.set_ugi(USER, null);
+    assertTokens(2, 1);
+
+    handlerProxy.get_table("db", "table");
+    assertTokens(1, 1);
+
+    handlerProxy.get_table("db", "table");
+    assertTokens(0, 1);
+
+    try {
+      handlerProxy.get_table("db", "table");
+      fail("Should have thrown exception.");
+    } catch (WaggleDanceServerException e) {
+      assertThat(e.getMessage(), is("[STATUS=429] Too many requests."));
+    }
+    
+    verify(client, times(3)).get_table("db", "table");
+    verify(client).set_ugi(USER, null);
+  }
+
+  @Test
+  public void testBucketExceptionStillDoCall() throws Exception {
+    Table table = new Table();
+    when(client.get_table("db", "table")).thenReturn(table);
+    BucketService mockedBucketService = Mockito.mock(BucketService.class);
+    when(mockedBucketService.getBucket(anyString())).thenThrow(new RuntimeException("Bucket exception"));
+    CloseableThriftHiveMetastoreIface proxy = new RateLimitingClientFactory(thriftClientFactory, mockedBucketService, bucketKeyGenerator)
+        .newInstance(metastore);
+
+    Table result = proxy.get_table("db", "table");
+    assertThat(result, is(table));
+  }
+
+  @Test
+  public void testInvocationHandlerThrowsCause() throws Exception {
+    when(client.get_table("db", "table")).thenThrow(new NoSuchObjectException("No such table"));
+    try {
+      handlerProxy.get_table("db", "table");
+      fail("Should have thrown exception.");
+    } catch (NoSuchObjectException e) {
+      assertThat(e.getMessage(), is("No such table"));
+    }
+  }
+  
+  @Test
+  public void testIgnoredMethods() throws Exception {
+    assertTokens(2, 2);
+    handlerProxy.set_ugi(USER, null);
```

Comment:
Considering that the list of ignored methods is fixed, shouldn't you include a test for each ignored method? We must ensure that those methods are always ignored.

Reply:
yeah I can split it up